### DISCUSSION TITLE
How to do one click file upload without a save or submit button

### DISCUSSION THREAD

**Discussion by abishekrsrikaanth** [livewire/livewire#2277] [1. Help]
Is there a way to perform a file upload with a single click without having a save or submit button?

I tried to use example for [File Validations](https://laravel-livewire.com/docs/2.x/file-uploads#file-validation), but I can't seem to get it to work. Here is the code I have. Any help is appreciated...

```
class Images extends Component
{
    use WithFileUploads;
    
    public $product;
    public $images;
    public $image;

    public function mount()
    {
        $this->images = $this->product->getMedia();
    }

    public function render()
    {
        return view('livewire.supplier.products.product.images');
    }

    public function updatedImage()
    {
        $this->validate([
            'image' => ['required', 'image'],
        ]);

        $this->product->addPhoto($this->image);

        $this->image = null;

        $this->images = $this->product->getMedia();
    }
}
```

Here is the blade code for the component

```
<input type="file" class="hidden" x-ref="image" wire:model="image">
<x-form.button type="button" x-on:click="$refs.image.click();" loading-target="image" additional-classes="ml-5" size="small">Upload Image</x-form.button>
```

**Comment by joshhanley** [comment]
@abishekrsrikaanth in your updatedImage method, you need to store the photo first and then add it to your product.

See this from the docs
> // Store in the "photos" directory in a configured "s3" bucket.
> $this->photo->store('photos', 's3');

So if you use the default public storage setup, and you want to save it to a photos folder, then you could just change 's3' to 'public'

$this->photo->store should then return the path of the saved file e.g. `photos/<some-hash-here>.png` which you can save to your product model and use the in the src attribute of your img tag.

Hope this helps!

**Reply by abishekrsrikaanth** [reply]
@joshhanley  @danharrin  The problem I am having is that the execution doesn't reach the first line of the method `updatedImage`, so if I try to upload a file that is not an image, it still goes through. I tried adding a `dd`, nothing happens. Not sure what is the issue. The custom button component doesn't seem to be the issue, because it doesn't have any issues with the rest of the app.

Also `$this->product->addPhoto($this->image);` is what saves the image to `s3`, the model actually handles saving the file to s3 with additional stuff happening. Doing it there because I use `spatie/laravel-medialibrary` to handle file uploads.

**Reply by joshhanley** [reply]
@abishekrsrikaanth can you share your blade view for the whole component? And also share your button component blade view? There must be something else going on, as I copied your snipped down and got it running without any problems.

Yeah I wouldn't worry about the storage just yet, if the updated hook isn't running theres either a problem with the Alpine click handler or the Livewire model binding, so we will try to get the updatedImage method firing first.

**Reply by Sindoweb** [reply]
I seem to have the same issue. The upload and temporary link all work, however Livewire is not running the function specified in wire:submit.prevent.

Blade:

`<form wire:submit.prevent="savehoofdautomaat" enctype="multipart/form-data">`
`<input type="file" accept="image/*" id="file1" class="hidden" wire:model="hoofdautomaat" />`
`</form>`

Component:

`public function savehoofdautomaat()
    {
	dd($this->step);
    }`

The script never reaches the dd()...

**Reply by joshhanley** [reply]
@Sindoweb try removing the enctype from the form element and seeing if it runs

**Reply by Sindoweb** [reply]
@joshhanley Unfortunately that doesn't change things. THe form on itself runs and images get stored on the server. It just won't go through the function as defined.

I'm starting to think this has to do with there not being a submit event. Similar to the 'required' attribute and forms being fired through ajax.

**Comment by danharrin** [comment]
Josh's answer sounds correct @abishekrsrikaanth, let us know how you get on!

Side note: the disk argument is optional, if you public disk is default, it will be saved there. If you're testing this, you'll want to set `FILESYSTEM_DRIVER` to `tmp-for-tests` in your `phpunit.xml`.

**Reply by boptom** [reply]
Is this documented? I just spent a day trying to work out why the examples weren't working.

**Reply by joshhanley** [reply]
@boptom which part are you asking whether it is documented exactly? Which examples weren't working?

**Reply by boptom** [reply]
@joshhanley The file uploads documentation here: https://livewire.laravel.com/docs/uploads mentions: The "upload" method is reserved

Then in the testing section: https://livewire.laravel.com/docs/uploads#testing-file-uploads

The code here uses the the "upload" call.
```
        Livewire::test(UploadPhoto::class)
            ->set('photo', $file)
            ->call('upload', 'uploaded-avatar.png');
```

Also, the Livewire component uses `saveAs`, with the `disk` parameter:
```
    public function upload($name)
    {
        $this->photo->storeAs('/', $name, disk: 'avatars');
    }
```

But `saveAs` doesn't have a `disk` parameter. It throws an "Unknown named parameter $disk" error.

-------

Other than those errors the testing is broken if default disk is used. For this component which doesn't specify a disk to save to:

```
<?php

namespace App\Livewire;

use Livewire\Attributes\Validate;
use Livewire\Component;
use Livewire\WithFileUploads;

class UploadPhoto extends Component
{
    use WithFileUploads;

    #[Validate('image|max:1024')] // 1MB Max
    public $photo;

    public $name = 'avatar.png';

    public function save()
    {
        $this->photo->storeAs('/', $this->name);
    }
}
```

the following test fails:
```
it('renders successfully fails', function () {
    Storage::fake();

    $file = UploadedFile::fake()->image('avatar.png');

    Livewire::test(UploadPhoto::class)
        ->set('photo', $file)
        ->set('name', 'uploaded-avatar.png')
        ->call('save');

    Storage::disk()->assertExists('uploaded-avatar.png');
});
```

with the error: 
```
  Unable to find a file or directory at path [uploaded-avatar.png].
```

However, if we use `tmp-for-tests` disk in the test:
```
it('renders successfully passes', function () {
    Storage::fake();

    $file = UploadedFile::fake()->image('avatar.png');

    Livewire::test(UploadPhoto::class)
        ->set('photo', $file)
        ->set('name', 'uploaded-avatar.png')
        ->call('save');

    Storage::disk('tmp-for-tests')->assertExists('uploaded-avatar.png');
});
```

It passes!

Another way around this is to use this setting in `phpunit.xml`
```
        <env name="FILESYSTEM_DISK" value="tmp-for-tests"/>
```

All of this is not in the Livewires docs AFAIK. A quick search for "tmp-for-tests" in the docs shows nothing.

---

I've created a minimal Laravel/Livewire repo with the issues here: https://github.com/boptom/livewire-file-test

Does this help? I can provide more details if required.

**Reply by joshhanley** [reply]
@boptom ah right yep, makes sense. PRs to fix the docs are welcome 🙂
